Rewite LocalExecutor to be simpler, and to shutdown cleanly on Python 3.10+

ashb opened the pull request:
Something changed between Python 3.7 and 3.10 meaning that a limited
parallelism LocalExecutor scheduler now doesn't shutdown cleanly on
receiving a signal.

On closer inspection of the limited vs unlimited path it apepars to me that
the code was "over-generalized" and the entire concept of `self.impl` has been
removed hopefully making this code much more direct and easier to understand.

The key things are now:

- When a task needs to be run, we send the message on a mp.SimpleQueue object,
  and increment an internal counter.

  (We use our own counter, not qsize method as that is not portable)

- Inside _check_workers we see if we think there are any outstanding messages,
  and create a worker if there are.

  The reason we do this is the on macOS (where the default mp start method is
  "spawn") a process will be started via `exeucte_async`, but it will take a
  second or two to pull the message of the queue, by which time the scheduler
  will have called `executor.sync()` again, meaning we'd over create workers
  (but never above the limit).

  Avoiding that case is why we keep the internal `_outstanding_messages`
  counter -- `self.activity_queue.empty()` would return False when the worker
  is booting up.

- Everytime the scheduler calls the `sync()` method we read out of the result
  queue and decrement the internal counter.

- We remove the entire use of `multiprocessing.Manager` -- it doesn't seem to
  do anything other than create queue objects but for our use it just adds
  complexity to understanding

- Almost as a side-effect we now only create worker subprocesses on demand,
  instead of pre-launching them.

  We do not currently shut down idle processes, though adding it would be
  quite straight forward if we wanted to in the future

This branch name was "rewrite-local-exexc-concurrentfutures" (sic) as when
originally opened in 2022 for 3.10 that was the plan.

However since then 3.12 has come out and it now starts issuing warnings when
Fork and threads are used, and concurrent.futures uses a thread internally, so
a different approach was used.

tanelk commented on airflow/executors/local_executor.py:
```diff
@@ -365,34 +154,42 @@ def execute_async(
     def terminate(self):
         """Terminate the executor is not doing anything."""
+        if self.futures_executor:
+            self.futures_executor.shutdown(wait=False)
```
Should we set `cancel_futures=True` in either of the shutdown calls.

tanelk commented on airflow/executors/local_executor.py:
```diff
@@ -201,161 +124,27 @@ class LocalExecutor(BaseExecutor):
-            else LocalExecutor.LimitedParallelism(self)
-        )
+        # This isn't _truely_ unlimited, but it should be good enough!
+        size = 2**30 if self.parallelism == 0 else self.parallelism
```
Interestingly there is no promises for the unlimited: https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#parallelism

ashb replied:
Huh. TIL

Sadly I think that although this isn't documented it's still "relied upon" behaviour so we can't just remove it.

ashb committed:
Rewite LocalExecutor to be simpler, and to shutdown cleanly on Python 3.10+

Something changed between Python 3.7 and 3.10 meaning that a limited
parallelism LocalExecutor scheduler now doesn't shutdown cleanly on
receiving a signal.

On closer inspection of the limited vs unlimited path it apepars to me that
the code was "over-generalized" and the entire concept of `self.impl` has been
removed hopefully making this code much more direct and easier to understand.

The key things are now:

- When a task needs to be run, we send the message on a mp.SimpleQueue object,
  and increment an internal counter.

  (We use our own counter, not qsize method as that is not portable)

- Inside _check_workers we see if we think there are any outstanding messages,
  and create a worker if there are.

  The reason we do this is the on macOS (where the default mp start method is
  "spawn") a process will be started via `exeucte_async`, but it will take a
  second or two to pull the message of the queue, by which time the scheduler
  will have called `executor.sync()` again, meaning we'd over create workers
  (but never above the limit).

  Avoiding that case is why we keep the internal `_outstanding_messages`
  counter -- `self.activity_queue.empty()` would return False when the worker
  is booting up.

- Everytime the scheduler calls the `sync()` method we read out of the result
  queue and decrement the internal counter.

- We remove the entire use of `multiprocessing.Manager` -- it doesn't seem to
  do anything other than create queue objects but for our use it just adds
  complexity to understanding

- Almost as a side-effect we now only create worker subprocesses on demand,
  instead of pre-launching them.

  We do not currently shut down idle processes, though adding it would be
  quite straight forward if we wanted to in the future

This branch name was "rewrite-local-exexc-concurrentfutures" (sic) as when
originally opened in 2022 for 3.10 that was the plan.

However since then 3.12 has come out and it now starts issuing warnings when
Fork and threads are used, and concurrent.futures uses a thread internally, so
a different approach was used.

ashb committed:
fixup! Rewite LocalExecutor to be simpler, and to shutdown cleanly on Python 3.10+

kaxil commented on airflow/executors/local_executor.py:
```diff
@@ -25,192 +25,139 @@
+    import signal
 
-    Executes the given command and puts the result into a result queue when done, terminating execution.
+    from setproctitle import setproctitle
```
This is not at the top for some reason, right? I remember discussing but can't remember

ashb replied:
on OSX importing secproctitle module does some stuff with threads.

ashb replied:
I'm not sure this actually makes any difference now I think about it, as this is in the mp process, but it's about to fork a new process anyway. I'll move it back.

ashb committed:
fixup! Rewite LocalExecutor to be simpler, and to shutdown cleanly on Python 3.10+

kaxil commented on airflow/executors/local_executor.py:
```diff
@@ -403,32 +202,91 @@ def execute_async(
+        # create too much
+
+        if self._outstanding_messages <= 0 or self.activity_queue.empty():
+            # Nothing to do, should we shut down idle workers?
```
Worth creating a GH issue for it if someone wants to take care of it

ashb replied:
Updated the comment. I don't think this is a big enough thing to create a GH issue for tbh

kaxil commented on airflow/executors/local_executor.py:
```diff
@@ -403,32 +202,91 @@ def execute_async(
-        self.impl.execute_async(key=key, command=command, queue=queue, executor_config=executor_config)
+        need_more_workers = len(self.workers) < self._outstanding_messages
+        if need_more_workers and (self.parallelism == 0 or len(self.workers) < self.parallelism):
+            self._spawn_worker()
```
If we have 2 workers and `_outstanding_messages` are 4, this is just going to create 1 worker. So we rely on multiple `_check_workers` invocations in sync/heartbeat. Worth adding a comment here or somewhere

ashb replied:
We call `_check_workers` directly from in `execute_async`, so we'll spawn workers directly after sending the message on the queue. I'll leave a comment to that effect.

kaxil commented on airflow/executors/local_executor.py:
```diff
@@ -403,32 +202,91 @@ def execute_async(
+        while not self.result_queue.empty():
+            key, state, exc = self.result_queue.get()
+            self._outstanding_messages = self._outstanding_messages - 1
+
```
Is there ever a risk of this becoming inconsistent due to exceptions or unexpected shutdowns?

ashb replied:
Good question, let me have a think.

ashb replied:
I think this is fine actually -- the only way we can decrement is if the work finishes to completion. If the process exits abnormally it by definition couldn't have sent the completion message.

So lets say that it picks up a message of the `activity_queue` and then dies.

We'll never get a completion message on `result_queue`, so `_outstanding_messages` will be off by 1.

Let me see if I can use a JoinableQueue (which has `task_done()` that we might be able to use instead. Maybe. It depends if we can get at the count without forcing a join.

ashb replied:
In 2f7d6c53d2 I've switched it to use a synchronized integer via `multiprocessing.Value` that we can pass to the worker process, so they can decrement the count as soon as it reads the message.

kaxil commented on airflow/executors/local_executor.py:
```diff
@@ -403,32 +202,91 @@ def execute_async(
+            self._outstanding_messages = self._outstanding_messages - 1
+
+            if exc:
+                # TODO: This needs a better stacktrace, it appears from here
+                if hasattr(exc, "add_note"):
+                    exc.add_note("(This stacktrace is incorrect -- the exception came from a subprocess)")
```
Planning to fix it here or separate PR?

ashb replied:
Planning to leave this -- it happens so rarely (as the `_execute_work` fns have try/except) so this is just an final line of defense to show something when the task goes wrong.

(The only way I was able to trigger this was a syntax error when developing it)

kaxil commented on tests/executors/test_local_executor.py:
```diff
@@ -44,85 +53,72 @@ def test_serve_logs_default_value(self):
-        fail_command = ["airflow", "tasks", "run", "failure", "some_parameter", "2020-10-07"]
-
         def fake_task_run(args):
+            print(repr(args))
```
```suggestion
```

ashb committed:
fixup! Rewite LocalExecutor to be simpler, and to shutdown cleanly on Python 3.10+

ashb committed:
fixup! Rewite LocalExecutor to be simpler, and to shutdown cleanly on Python 3.10+
